### ISSUE TITLE
Not Working in Firefox

### ISSUE THREAD

**Issue by larry-beck** [browserpass/browserpass-otp#2]
Also, seems to work right in Chromium, but not at all in Firefox. Also no extension icon showing as it does in Chromium after installing the OTP extension.

**Comment by erayd** [comment]
@larry-beck No extension icon is expected, as this extension does not yet have a UI of any kind.

Could you please clarify which version of Browserpass you have installed in Firefox? Firefox can lag somewhat with its extension updates, and you need at least v3.0.12 of Browserpass for this to function.

**Comment by larry-beck** [comment]
I have 3.0.12 installed manually and 0.1.0 of OTP, but it was manually when first released and they were not showing on Mozilla Addon store, I have since installed release 3.0.12 and 0.1.0 of them from online release and it still does not auto copy OTP for website to clipboard. When I do the same thing in Chromium, it works fine, copies OTP for website to clipboard and I can paste the OTP fine.

**Comment by erayd** [comment]
Which version of Firefox are you using, and on which OS?

Can you explain the steps you are taking prior to attempting to paste the OTP code? When you do try to paste it, what gets pasted instead?

I can confirm it's working correctly in Firefox 66.0.2 on Gentoo, so if we can pin down what's different about your environment, it will help me to figure out what might be going on.

**Comment by larry-beck** [comment]
1. Firefox 66.0.3 on macOS 10.14.4
2. Go to website login page with username/pass and OTP saved in Passwordstore
3. Click browserpass button in Firefox, it fills the username/pass correctly
4. The website screen changes to OTP entry page, I click paste via OS keystrokes or Edit paste menu and an old copy contents is pasted from a previous copy and paste routine.
5. This same procedure works fine with Chromium 73.0.3683.103 on same OS.

**Comment by erayd** [comment]
Hmmm. Do the copy user / copy pass functions of browserpass work correctly in Firefox for you?

If you invoke the browserpass entry with the debug console of browserpass-otp open, do you see anything in the console?

**Comment by DamienCassou** [comment]
It works for me:

- Firefox 66.0.2
- Fedora 29

**Comment by larry-beck** [comment]
> Hmmm. Do the copy user / copy pass functions of browserpass work correctly in Firefox for you?

No copy/paste of username and/or password do not work either in Firefox, work fine in Chromium on same computer.

> If you invoke the browserpass entry with the debug console of browserpass-otp open, do you see anything in the console?

Yes, the following error:
03:01:01.247 Promise resolved after context unloaded
settings.js:436
    <anonymous> moz-extension://5457c985-12c1-544a-89ad-bb6a02552296/lib/settings.js:436

**Comment by larry-beck** [comment]
Seems the OTP password copy is working now with latest OTP release in Mozilla Store, but you cannot click the username or password button in the extension popup and copy/paste either one, the only thing that pastes is the OTP for that pass entry

**Comment by larry-beck** [comment]
> Seems the OTP password copy is working now with latest OTP release in Mozilla Store, but you cannot click the username or password button in the extension popup and copy/paste either one, the only thing that pastes is the OTP for that pass entry

That is only partially true, it works as stated above in Firefox Beta, but not in Firefox release. I have removed and reinstalled browserpass 3.0.12 and browserpass-otp 0.1.0 in Release Firefox, still does not work there, but copy/paste of OTP in Firefox Beta works fine, but you cannot copy/paste username or password for site from buttons in extension popup.

**Comment by erayd** [comment]
> Seems the OTP password copy is working now with latest OTP release in Mozilla Store, but you cannot click the username or password button in the extension popup and copy/paste either one, the only thing that pastes is the OTP for that pass entry.

That's expected - see bug #1.

Firefox 66 should work just fine, and I still haven't managed to reproduce your issue. How are you installing both extensions?

**Comment by larry-beck** [comment]
Again, in Firefox beta v67.0b11, OTP works fine, but not in Firefox release v66.0.3, OTP does not work. I have uninstalled and reinstalled both extensions from addons.mozilla.org, Browserpass v3.0.12 and Browserpass OTP v0.1.0. Works in Firefox v67, not in Firefox v66, tell me what else you want to know and I will provide info from my side.

**Comment by erayd** [comment]
Noting you're having issues with both extensions, let's take remove browserpass-otp from consideration for now. They both use the same method of pushing data to the clipboard.

@maximbaz - do you have any ideas? Browserpass clipboard functionality in Firefox 66 on MacOS apparently doesn't work at all here, but does work on other platforms. Firefox 67 beta works on MacOS.

Do we have any other Mac users who might be able to confirm this?

Right now I'm out of ideas - perhaps this is a Firefox bug?

**Comment by larry-beck** [comment]
Gents, let me recap what is working and not working and where for me on macOS v10.14.4

I have the following installed for each Firefox noted below, both extensions installed directly from addons.mozilla.org, nothing direct from Github. I have Browserpass v3.0.12 and Browserpass OTP v0.1.0

1. **Firefox v66.0.3** - **Browserpass** works fine for auto filling username/password on website, however if I try to manually copy username or password from buttons in extension, it does not work for either of them. **Browserpass OTP** does not work as when I select the entry for the website from the extension, it should copy the OTP that is setup for the entry in passwordstore to the clipboard as it does in Chromium version, but it does not copy anything to the clipboard, when I try to paste, macOS still has the same information from my last system wide copy operation.

2. **Firefox v67.0b11** - **Browserpass** works fine for auto filling username/password on website, however if I try to manually copy username or password from buttons in extension, it does not work for either of them. **Browserpass OTP** does work here as expected so when I select an entry to auto fill the website with username/password, it copies the OTP for that entry to the clipboard and when I paste, I get my OTP for the website.

Any other questions, fire away and i can provide feedback or more information about my install and setup.

**Comment by erayd** [comment]
@larry-beck (2) is expected behavior. The only issue there is bug #1, which is a known conflict. This will be fixed.

(1) is a problem, and right now I have no idea why it's happening, and I can't reproduce it on Linux (also Firefox v66). Noting it works for you in v67, I'm wondering if this might actually be a Firefox bug, rather than a problem with browserpass, but I don't actually know. I'm hoping that @maximbaz might have an idea of what might be happening, because right now, I don't.

Do you know anyone else who uses MacOS and might be able to test?

**Comment by maximbaz** [comment]
this is very suspicious. @larry-beck, when you are testing clipboard functionality in browserpass itself, are you testing with otp extension removed? because due to #1 it is not supposed to work right now anyway. 

Maybe we should revisit this after #1 is fixed, in case this was caused by some sort of race condition...

**Comment by larry-beck** [comment]
With 3.0.13 and 0.2.0 of both extensions, now all is working fine in Firefox beta v67 as well as Chromium on macOS. OTP shows code and copy ability, Browserpass copies username and password when selecting icon in extension as well.

However, with 3.0.13 and 0.2.0, in Firefox v66 public version, OTP ext. shows code and you can copy it, however, copying username or password for site still does not work, I know about some other bug, I am just telling what is working with both exts. on my setup of macOS.

**Comment by maximbaz** [comment]
We should strive to fix it, but given that we can't reproduce this and are out of ideas, and given that it magically works in the next version of Firefox, I have some doubts that we will be able to do anything about this issue. If anyone has ideas, please share.

But it's not all bad, Firefox 67 is going to be released in less than a month, on May 14, so at least you are guaranteed to have a fix relatively soon, one way or another 😉

**Comment by larry-beck** [comment]
All seems fine, I tested on another Mac with Firefox release, no issues at all, copy/paste username and password as well as OTP works as expected. I have something restricted in my Firefox release install v66 as I have it heavily privacy locked down, so something turned on/off in Firefox settings is making it not work for copy/paste in my v66, while my v67 beta copy is vanilla, nothing special turned on/off in advanced settings of Firefox. I am closing this now as all issues have been resolved to my knowledge.
